homeassistant/components/zha/__init__.py

```diff
@@ -369,6 +419,122 @@ def zdo_command(self, tsn, command_id, args):
+    Stateless sensors such as remotes are expected to generate an event
+    instead of a sensor entity in hass.
+    """
+
+    def __init__(self, hass, cluster, domain, discovery_info):
+        """Register callback that will be used for signals."""
+        self._hass = hass
+        self._cluster = cluster
+        self._cluster.add_listener(self)
+        ieee = discovery_info['endpoint'].device.ieee
+        ieeetail = ''.join(['%02x' % (o, ) for o in ieee[-4:]])
+        if discovery_info['manufacturer'] and discovery_info['model'] is not \
+                None:
+            self._id = "{}.{}_{}_{}_{}{}".format(
+                domain,
+                slugify(discovery_info['manufacturer']),
+                slugify(discovery_info['model']),
+                ieeetail,
+                discovery_info['endpoint'].endpoint_id,
+                discovery_info.get('entity_suffix', '')
+            )
+        else:
+            self._id = "{}.zha_{}_{}{}".format(
+                domain,
+                ieeetail,
+                discovery_info['endpoint'].endpoint_id,
+                discovery_info.get('entity_suffix', '')
+            )
+
+
+class ZHASwitchEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0040):
+            self._hass.bus.fire(
+                OFF_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id in (0x0001, 0x0041, 0x0042):
+            self._hass.bus.fire(
+                ON_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+        elif command_id == 0x0002:
+            self._hass.bus.fire(
+                TOGGLE_EVENT_KEY,
+                {DEVICE: self._id},
+                EventOrigin.remote
+            )
+
+
+class ZHALevelEvent(ZHAEvent):
+    """Switch / remote event for zha"""
+
+    def __init__(self, hass, cluster, discovery_info):
+        """Initialize Switch."""
+        super().__init__(hass, cluster, SWITCH, discovery_info)
+        self._level = 0
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id in (0x0000, 0x0004):  # move_to_level, -with_on_off
+            self.set_level(args[0])
+        elif command_id in (0x0001, 0x0005):  # move, -with_on_off
+            # We should dim slowly -- for now, just step once
+            rate = args[1]
+            if args[0] == 0xff:
+                rate = 10  # Should read default move rate
+            self.move_level(-rate if args[0] else rate)
+        elif command_id == 0x0002:  # step
```

Comment:
there's 'step' and 'step_with_on_off' == '0x06. We should handle both of those. Check #14756

Reply:
Will do, ty